doc/topics/async-system.md

```diff
@@ -0,0 +1,209 @@
+
+* _Resolved_: The operation has completed successfully and the result value is available.
+* _Rejected_: The operation failed, and an exception describing the error is available.
+
+If we want to block the current thread and wait for the request to complete, we can call [waitInMainThread](@ref CesiumAsync::Future::waitInMainThread):
+
+\snippet{trimleft} ExamplesAsyncSystem.cpp wait-in-main-thread
+
+`waitInMainThread` blocks until the `Future` completes. If it is resolved, `waitInMainThread` returns its value. If it is rejected, `waitInMainThread` throws an exception representing the error.
+
+> [!note]
+> The `std::move` is needed because `Future<T>` is meant to be consumed just once. This allows for best efficiency because the value resulting from the asynchronous operation can be provided to the caller without ever making a copy of it.
+
+In practice, though, we almost never call `waitInMainThread`. After all, what good is it to start an asynchronous operation if we're just going to block the calling thread waiting for it? Instead, we register a _continuation function_ with the `Future`.
+
+## Continuation Functions {#continuation-functions}
+
+A continuation function is a callback - usually a lambda - that is invoked when the `Future` resolves. The name comes from the idea that the execution "continues" with that function after the async operation. Continuations are registered with the `then...` methods on the `Future`. Different `then` functions are used to control _in what thread_ the callback is invoked:
+
+* [thenInMainThread](@ref CesiumAsync::Future::thenInMainThread): The continuation function is invoked in the "main" thread. Many applications have a clear idea of which thread they consider the main one, but in Cesium Native, "main thread" very specifically means the thread that called `waitInMainThread` or [dispatchMainThreadTasks](@ref CesiumAsync::AsyncSystem::dispatchMainThreadTasks) on the `AsyncSystem`. This need not be the same as the thread your application considers to be the main one.
+* [thenInWorkerThread](@ref CesiumAsync::Future::thenInWorkerThread): The continuation function is invoked in a background worker thread by calling [startTask](@ref CesiumAsync::ITaskProcessor::startTask) on the `ITaskProcessor` instance with which the `AsyncSystem` was constructed.
+* [thenInThreadPool](@ref CesiumAsync::Future::thenInThreadPool): The continuation function is invoked in a specified [ThreadPool](@ref CesiumAsync::ThreadPool).
+* [thenImmediately](@ref CesiumAsync::Future::thenImmediately): The continuation function is invoked immediately in whatever thread happened to resolve the `Future`. This is only appropriate for work that completes very quickly and is safe to run in any thread.
```

Comment:
"This is only appropriate for work that completes very quickly..." is that actually true if the future is known to be resolved in a worker thread e.g. we've written `thenInWorkerThread()` in the continuation chain?

Reply:
Yes, even then. `thenInWorkerThread(...).thenImmediately(...)` is identical to `thenInWorkerThread(...).thenInWorkerThread(...)` in terms of where the second continuation runs. But the latter is clearer and won't accidentally switch to running in a different thread if someone inserts a `thenInMainThread` in between the two.

Reply:
Do you mean "the former is clearer?"  If these two forms are identical, then what's the way to schedule new, expensive work?

Reply:
No, I meant that `thenInWorkerThread(...).thenInWorkerThread(...)` is the clearer of the two, because it makes it clear that both lambdas are run in a worker thread (and not, for example, taking up time on the main thread).

> If these two forms are identical, then what's the way to schedule new, expensive work?

This _is_ the way to schedule new, expensive work, despite the fact that they're identical. Maybe I'm not understanding your concern?

Ok, so let me try to explain the nuance here... When a Future resolves in a worker thread (for example, the continuation passed to the first `thenInWorkerThread` in our current example returns), and its continuation is registered with `thenInWorkerThread`, that continuation will run _immediately_ in the same worker thread that just resolved the Future. Exactly the same as if that continuation were registered with `thenImmediately`.

However, if the Future resolved in some other thread (not a worker thread), and its continuation is registered with `thenInWorkerThread`, then the continuation will be _scheduled_ to run in a worker thread. It will _not_ run immediately.

This is an optimization, to avoid the overhead of repeatedly scheduling work that should run in the same context without requiring us to use explicit `thenImmediately` calls.

`thenImmediately` is error prone because we don't know for sure which thread resolves a future in all cases, so it's dangerous to make assumptions about it. For example, if I call `IAssetAccessor::get`, what thread resolves that Future? In Unreal, it happens to be in the main thread. In other applications, it might be a worker or an HTTP thread or something like that. We can't know for sure. But by simply using the correct `then..` for the context I want my continuation to run, I get the most efficient performance if that place happens to be the same context the Future was resolved. If I truly don't care where it runs - because it's quick and thread agnostic - and want to minimize scheduling overhead in all cases, _then_ it is appropriate to use `thenImmediately`.

The only downside I can think of to this has to do with prioritization. If I have `thenInWorkerThread(...).thenInWorkerThread(...)`, should other work that was scheduled earlier get to run in between these two worker thread lambdas? With the current implementation, it never will. I don't think this is a big deal. An early version of AsyncSystem gave users control over this by calling something like `thenInWorkerThread` versus `thenImmediatelyInWorkerThread`, but it was just a lot of unnecessary complexity IMO and that's why it's gone now.

Reply:
I think I understand now. I was fixated on a chain of continuation functions where you know explicitly what kind of thread will resolve the continuation because... you just wrote `thenInWorkerThread` or `thenInMainThread`. Like you say, if the future is returned from a function, or if the surrounding code changes in the future, you just wouldn't know.